### ISSUE TITLE
Mono<T[]> Mono#when(Iterable<? extends Mono<? extends T>>) failed with ClassCastException

### ISSUE THREAD

**Issue by raymondhjm** [reactor/reactor-core#123]
The underlying code snippet will failed with `java.lang.ClassCastException`, this happens at 3.0.0.RC1.

When I call

```java
Mono<String[]> mono = Mono.when(new Mono[]{Mono.just("hello"), Mono.just("world")});
mono.subscribe(results -> {});
```

I got

```java
reactor.core.Exceptions$BubblingException: java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Ljava.lang.String;

	at reactor.core.Exceptions.bubble(Exceptions.java:95)
	at reactor.core.Exceptions.onErrorDropped(Exceptions.java:199)
	at reactor.core.publisher.LambdaSubscriber.onError(LambdaSubscriber.java:126)
	at reactor.core.publisher.LambdaSubscriber.onNext(LambdaSubscriber.java:145)
	at reactor.core.publisher.Operators$DeferredScalarSubscriber.complete(Operators.java:797)
	at reactor.core.publisher.MonoWhen$MonoWhenCoordinator.signal(MonoWhen.java:171)
	at reactor.core.publisher.MonoWhen$MonoWhenSubscriber.onNext(MonoWhen.java:215)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:1293)
	at reactor.core.publisher.MonoWhen$MonoWhenSubscriber.onSubscribe(MonoWhen.java:205)
	at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:49)
	at reactor.core.publisher.MonoWhen$MonoWhenCoordinator.subscribe(MonoWhen.java:109)
	at reactor.core.publisher.MonoWhen.subscribe(MonoWhen.java:81)
	at reactor.core.publisher.Mono.subscribeWith(Mono.java:2243)
	at reactor.core.publisher.Mono.subscribe(Mono.java:2205)
	at reactor.core.publisher.Mono.subscribe(Mono.java:2164)
Caused by: java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Ljava.lang.String;
	at reactor.core.publisher.LambdaSubscriber.doNext(LambdaSubscriber.java:155)
	at reactor.core.publisher.LambdaSubscriber.onNext(LambdaSubscriber.java:136)
```

This is caused by having an `Object[]` in `reactor.core.publisher.MonoWhen$MonoWhenCoordinator.signal()` to accumulate the value of type-parameterized `Iterable<Mono<T>>`. In this case, an `Object[]` is used to store element with `java.lang.String` type, and finally class cast to `String[]` at runtime. Casting `String[]` to `Object[]` is OK because the Class of `Object[]` is super of the Class of `String[]`, but not the case inverse.

Unfortunely, due to the limitation of Java generic type erasing at runtime, we're unable to instantiate a generic type array, and `complete((T[])o)` will pass compiler checking at `reactor.core.publisher.MonoWhen$MonoWhenCoordinator.signal(MonoWhen.java:171)` because compiler will treat as `complete((Object[])o)`, but fail eventually.

**Comment by smaldini** [comment]
This one was a critical miss, thanks for detecting that one up - we had to change the signature slightly but given the method was hard to work with and usually the T1, T2 alternatives take over, I expect this to be a welcome fix :)

**Comment by raymondhjm** [comment]
Hi @smaldini , thanks for the quick fixing.

I have been searching for such an async library for a long time to use as the base stone acrossing my whole application, with the service layer, data accessing layer as well as the cooperating downstreams, finally I came to project reactor.

I ran into a problem when I played with the new API `Mono<Void> when(final Iterable<? extends Publisher<Void>>)`.

Here's one of the daily usage scenarios:

One of my service depends on an unpredictable length of `Mono` list (I can limit the upper length bound) which were backed by async I/O to deal with different downstream system, the `when(Iterable<Mono>)` is a perfect case for this situation, it gives a "joining/resuming" point to assemble results from vary downstream systems in the future and resuming deferred handling.

A simple method to transform 'List of String Mono' to a 'Mono of String List'
```Java
Mono<List<String>> listToMono(List<Mono<String>> asyncMonos) {
    List<Mono<Void>> voidMonos = StreamSupport.stream(asyncMonos.spliterator(), false).map(Mono::then).collect(Collectors.toList());
    return Mono.when(voidMonos).then(result -> {
        List<String> stringList = new ArrayList<>();
        asyncMonos.forEach(mono -> stringList.add(mono.block()));
        return Mono.just(stringList);
    });
}
```

The following simple usage is expected to output "hello,world", but finally shows nothing.
```Java
listToMono(Arrays.asList(Mono.just("hello"), Mono.just("world")))
                .subscribe(list -> String.join(",", list));
```

The problem is that the `Mono<Void> when(final Iterable<? extends Publisher<Void>>)` will not publish the `next` event, but only a single `complete` event instead (due to the `VOID_FUNCTION` zipper at `MonoWhenCoordinator#signal()`??, not quite sure).

Here's the tweak to get the expected output “hello,world":

```Java
List<Mono<String>> monoList = Arrays.asList(Mono.just("hello"), Mono.just("world"));
listToMono(monoList)
    .subscribe(__ -> {}
        Throwable::printStackTrace,
        () -> {
            List<String> list = monoList.stream().map(mono -> mono.block()).collect(Collectors.toList());
            System.out.println(String.join(",", list));
        });
```

However, this usage leads to several side effects:

1. It breaks the origin contract of `Mono.then()`, the `next` event never publish, the lambda will never get called, thus breaks the whole reactive pipeline.
2. The only way to collect result in by listening on the `completion` event.
  * It's not intuitive and nature, one had to work around to get result :(
  * It requires transitive access to the original `List<Mono>>`, which introduces unnecessary complexity and leads to bad code design.
  * There's no any other way except `block()` method to access `Mono` value, this is not the recommend usage to `block()`, as it's name implied "blocking access"

Since we had to call `block()`, another tricky situation happens...

Consider following usage:

```Java
AtomicInteger counter = new AtomicInteger(0);
List<Mono<String>> monoList = Arrays.asList(Mono.fromCallable(() -> String.valueOf("hello" + counter.incrementAndGet())));
listToMono(monoList)
    .subscribe(__ -> {}
        Throwable::printStackTrace,
        () -> {
            List<String> list = monoList.stream().map(mono -> mono.block()).collect(Collectors.toList());
            System.out.println(String.join(",", list));
        });
```

What we're expecting "hello1", but finally outputs "hello2"...

As you may guessed, the callable Mono has been called twice, one is to initiate the publisher data source, one is initiated by manually calling `block()` in  the `completion` event handling.

However I found another tweak to handle this problem by turning the source `Mono` with a `cache()` call.

This is the final version, pay attention to the **cache()** call.

```Java
AtomicInteger counter = new AtomicInteger(0);
List<Mono<String>> monoList = Arrays.asList(Mono.fromCallable(() -> String.valueOf("hello" + counter.incrementAndGet())).cache());
listToMono(monoList)
    .subscribe(__ -> {}
        Throwable::printStackTrace,
        () -> {
            List<String> list = monoList.stream().map(mono -> mono.block()).collect(Collectors.toList());
            System.out.println(String.join(",", list));
        });
```

This will correctly prints out "hello1". I'm sure this increases overhead.

Usability is a critical consideration to an API design, but I think this won't make anyone happy, at least I think we need to find a better way to deal with this.

Let's reconsider this API and come back to earlier, have you considered changing the return type simply from `Mono<T[]>` to `Mono<List<T>>`? With generic List<T> type, we could utilise compiler's type checking and easily zip the result value to the list, as well as retaining all the semantic that T[] could provide.

Hope this helps!

**Comment by smaldini** [comment]
You can still do that with when(Function, Iterable/Vararg) or am I wrong ?

**Comment by raymondhjm** [comment]
I missed the newly added when(Function, Iterable/Vararg) method, too bad.

I played with it, it solves the problem I listed above, that's great. With one more issue, the Function receives Object[]. As we know it's difficult to deal with generic type array, maybe we need accept the trade-off.

Another question, what's the different with zip(Function, Iterable/Vararg)? I can see zip() provides stronger compiler checking with generic type "T" as when() provides wildcard matching "?". The behavior are basically identical?
